File: apps/zxtune-android/zxtune/src/main/java/app/zxtune/fs/modland/Catalog.java
/**
 * @file
 * @brief Catalog interface
 * @author vitamin.caig@gmail.com
 */

package app.zxtune.fs.modland;

import android.content.Context;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import java.io.IOException;
import java.nio.ByteBuffer;

import app.zxtune.fs.HttpProvider;
import app.zxtune.fs.cache.CacheDir;

public abstract class Catalog {

  public abstract static class GroupsVisitor {

    public void setCountHint(int size) {
    }

    public abstract void accept(Group obj);
  }

  public abstract static class TracksVisitor {

    public void setCountHint(int size) {
    }

    //too many tracks possible, so enable breaking
    public abstract boolean accept(Track obj);
  }

  public interface Grouping {

    /**
     * Query set of group objects by filter
     * @param filter letter(s) or '#' for non-letter entries
     * @param visitor result receiver
     */
    void query(String filter, GroupsVisitor visitor) throws IOException;

    /**
     * Query single group object
     * @param id object identifier
     * @return null if no object found
     */
    @Nullable
    Group query(int id) throws IOException;

    /**
     * Query group's tracks
     * @param id object identifier
     * @param visitor result receiver
     */
    void queryTracks(int id, TracksVisitor visitor) throws IOException;

    /**
     * Query track by name
     * @param id object identifier
     * @param filename track filename
     * @return null if nothing found
     */
    @Nullable
    Track findTrack(int id, String filename) throws IOException;
  }

  public abstract Grouping getAuthors();

  public abstract Grouping getCollections();

  public abstract Grouping getFormats();

  /**
   * Get track file content
   * @param path path to module starting from /pub/..
   * @return content
   */
  @NonNull
  public abstract ByteBuffer getTrackContent(String path) throws IOException;

  public static Catalog create(Context context, HttpProvider http, CacheDir cache) throws IOException {
    final Catalog remote = new RemoteCatalog(http);
    final Database db = new Database(context);
    return new CachingCatalog(remote, db, cache);
  }
}

File: apps/zxtune-android/zxtune/src/main/java/app/zxtune/fs/amp/Catalog.kt
/**
 * @file
 * @brief Catalog interface
 * @author vitamin.caig@gmail.com
 */

package app.zxtune.fs.amp

import android.content.Context
import app.zxtune.fs.http.MultisourceHttpProvider
import java.io.IOException

interface Catalog {

    abstract class WithCountHint {
        open fun setCountHint(count: Int) {
        }
    }

    abstract class GroupsVisitor : WithCountHint() {
        abstract fun accept(obj: Group)
    }

    abstract class AuthorsVisitor : WithCountHint() {
        abstract fun accept(obj: Author)
    }

    abstract class TracksVisitor : WithCountHint() {
        abstract fun accept(obj: Track)
    }

    abstract class FoundTracksVisitor : WithCountHint() {
        abstract fun accept(author: Author, track: Track)
    }

    /**
     * Query all groups
     * @param visitor result receiver
     */
    @Throws(IOException::class)
    fun queryGroups(visitor: GroupsVisitor)

    /**
     * Query authors by handle filter
     * @param handleFilter letter(s) or '0-9' for non-letter entries
     * @param visitor result receiver
     */
    @Throws(IOException::class)
    fun queryAuthors(handleFilter: String, visitor: AuthorsVisitor)

    /**
     * Query authors by country id
     * @param country scope
     * @param visitor result receiver
     */
    @Throws(IOException::class)
    fun queryAuthors(country: Country, visitor: AuthorsVisitor)

    /**
     * Query authors by group id
     * @param group scope
     * @param visitor result receiver
     */
    @Throws(IOException::class)
    fun queryAuthors(group: Group, visitor: AuthorsVisitor)

    /**
     * Query authors's tracks
     * @param author scope
     * @param visitor result receiver
     */
    @Throws(IOException::class)
    fun queryTracks(author: Author, visitor: TracksVisitor)

    /**
     * Find tracks by query substring
     * @param query string to search in filename/title
     * @param visitor result receiver
     */
    @Throws(IOException::class)
    fun findTracks(query: String, visitor: FoundTracksVisitor)

    companion object {
        const val NON_LETTER_FILTER = "0-9"

        @JvmStatic
        fun create(context: Context, http: MultisourceHttpProvider): CachingCatalog {
            val remote = RemoteCatalog(http)
            val db = Database(context)
            return CachingCatalog(remote, db)
        }
    }
}
